parachain/pallets/basic-channel/src/outbound/mod.rs

```diff
@@ -215,56 +271,128 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// For every account id in the enqueued messages, create an Eth ABI-encoded message
+		/// bundle containing the messages for that account. Hash the ethabi encoding of these
+		/// message bundles with a Merkle tree to produce a commitment hash. Then store the
+		/// commitment on the parachain, emit an event with the commitment and SCALE-encoded
+		/// message bundles and persist the ethabi-encoded message bundles to off-chain storage.
 		fn commit() -> Weight {
-			let messages = <MessageQueue<T>>::take();
-			if messages.is_empty() {
+			// TODO: consider using mutate_exists here. If some part of emitting message bundles
+			// fails, we don't want the MessageQueue to be empty.
+			let message_queue = <MessageQueue<T>>::take();
+			if message_queue.is_empty() {
 				return T::WeightInfo::on_initialize_no_messages();
 			}
 
-			let nonce = <Nonce<T>>::get();
-			let next_nonce = nonce.saturating_add(1);
-			<Nonce<T>>::put(next_nonce);
+			// Store these for the on_initialize call at the end
+			let message_count = message_queue.len() as u32;
+			let average_payload_size = Self::average_payload_size(&message_queue);
 
-			let bundle = MessageBundle {
-				source_channel_id: ChannelId::Basic as u8,
-				nonce: next_nonce,
-				messages: messages.clone(),
-			};
+			let message_bundles = Self::make_message_bundles(message_queue);
+			let eth_message_bundles: Vec<Vec<u8>> = message_bundles
+				.clone()
+				.into_iter()
+				.map(|bundle| ethabi::encode(&vec![bundle.into()]))
+				.collect();
+
+			// TODO: Maybe undo the O type param in merkle_root, since we can convert between H256
+			// and [u8; 32] easily and the merkle_root implementation mentioned that [u8; 32] was
+			// chosen for "a more optimised implementation".
+			let commitment_hash = merkle_root::<
+				<T as Config>::Hashing,
+				Vec<Vec<u8>>,
+				Vec<u8>,
+				<<T as Config>::Hashing as Hash>::Output,
+			>(eth_message_bundles.clone());
 
-			let commitment_hash = Self::make_commitment_hash(&bundle);
 			let digest_item =
 				AuxiliaryDigestItem::Commitment(ChannelId::Basic, commitment_hash.clone()).into();
 			<frame_system::Pallet<T>>::deposit_log(digest_item);
-			Self::deposit_event(Event::Committed { hash: commitment_hash, data: bundle });
 
-			T::WeightInfo::on_initialize(
-				messages.len() as u32,
-				Self::average_payload_size(&messages),
-			)
+			Self::deposit_event(Event::Committed {
+				hash: commitment_hash,
+				data: message_bundles.clone(),
+			});
+
+			// Persist ABI-encoded leaves to off-chain storage
+			let tokens = eth_message_bundles
+				.into_iter()
+				.map(|bundle| Token::Bytes(bundle))
+				.collect::<Vec<Token>>();
+			set(commitment_hash.as_bytes(), &ethabi::encode(&tokens));
```

Comment:
While the individual leaves should be ABI-encoded, the container in which they are serialized should be SCALE-encoded for efficiency.

For example, you could persist this SCALE-encoded container to offchain storage. Each individual `Vec<u8>` still represents an ABI-encoded leaf.
```rust
#[derive(Encode, Decode)]
struct StoredLeaves(Vec<Vec<u8>>)
```